### ISSUE TITLE
Data type compatibility issue with SPARQL between VOS 6.1.6 and 6.1.8

### ISSUE THREAD

**Issue by fgiasson** [openlink/virtuoso-opensource#131]
A compatibility issue has been found regarding datatype handling in a specific SPARQL query with VOS 6.1.8.

This issue has been confirmed by Hugh.

I did compile VOS 6.1.8 without any issues, then I ran it with a VOS 6.1.6 database file. There was no issues, and everything was working as expected, until I use the following SPARQL query:

``` sql
select ?s ?p ?o (DATATYPE(?o)) as ?otype (LANG(?o)) as ?olang
where
{
  {
    select distinct ?s from <http://localhost/datasets/organizations/import>
    where
    {
      ?s a ?type.
    }
    limit 50
    offset 0
  }

  ?s ?p ?o
}
```

If I am requesting the following mimes to the /sparql/ endpoint:
1. application/sparql-results+xml
2. application/sparql-results+json

I am getting the following content from the endpoint:

``` xml
<sparql xmlns="http://www.w3.org/2005/sparql-results#" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.w3.org/2001/sw/DataAccess/rf1/result2.xsd">
 <head>
  <variable name="s"/>
  <variable name="p"/>
  <variable name="o"/>
  <variable name="otype"/>
  <variable name="olang"/>
 </head>
 <results distinct="false" ordered="true">
  <result>
   <binding name="s"><uri>http://localhost/resources/organization/Random_House</uri></binding>
   <binding name="p"><uri>http://www.w3.org/1999/02/22-rdf-syntax-ns#type</uri></binding>
   <binding name="o"><uri>http://dbpedia.org/ontology/Publisher</uri></binding>
  </result>
  <result>
   <binding name="s"><uri>http://localhost/resources/organization/Random_House</uri></binding>
   <binding name="p"><uri>http://www.w3.org/1999/02/22-rdf-syntax-ns#type</uri></binding>
   <binding name="o"><uri>http://www.w3.org/ns/org#Organization</uri></binding>
  </result>
  <result>
   <binding name="s"><uri>http://localhost/resources/organization/Random_House</uri></binding>
   <binding name="p"><uri>http://xmlns.com/foaf/0.1/homepage</uri></binding>
   <binding name="o"><uri>http://www.randomhouse.com</uri></binding>
  </result>
  <result>
   <binding name="s"><uri>http://localhost/resources/organization/Random_House</uri></binding>
   <binding name="p"><uri>http://dbpedia.org/ontology/numberOfEmployees</uri></binding>
   <binding name="o"><literal datatype="http://www.w3.org/2001/XMLSchema#integer">5712</literal></binding>Virtuoso 22023 Error SR544: Function __xsd_type() can not find XML Schema datatype that matches SQL datatype UNAME (217)

SPARQL query:
select ?s ?p ?o (DATATYPE(?o)) as ?otype (LANG(?o)) as ?olang
      where
      {
        {
          select distinct ?s from <http://localhost/datasets/organizations/import>
          where
          {
            ?s a ?type.
          }
          limit 50
          offset 0
        }

        ?s ?p ?o
      }
```

Obviously, the error is related to the datatype request. However, I am not sure why it chokes there, and in that way. If I request the text/html mime, it works fine.

**Comment by markwilkinson** [comment]
Oddly, when I hit this error a few minutes ago on my own datasets, I discovered that if you "group by ?o", the problem goes away!

**Comment by markwilkinson** [comment]
...and that solution just stopped working...

**Comment by HughWilliams** [comment]
In what way has it stopped working, what had changed in your environment to cause this ?

**Comment by markwilkinson** [comment]
Unfortunately, it's not _my_ environment.  I am querying a third-party endpoint.  It's possible that they have changed something in their environment in the past few days.

Symptoms:  A piece of code that was working last week stopped working today.  I checked, and it was because one of my queries was triggering the error described above.

BACKGROUND: When I first wrote that script, I experienced the same error mentioned above, but when I added the "group by" clause to my query the error disappeared and my query worked for many weeks.  ...until today... 

I'm afraid that's the most I can report.  Sorry!

**Comment by HughWilliams** [comment]
what is the 3rd party endpoint you are querying , http://s4.semanticscience.org:14001/sparql  ?

**Comment by markwilkinson** [comment]
correct

On 05/13/2014 06:09 PM, HughWilliams wrote:

> what is the 3rd party endpoint you are querying , 
> http://s4.semanticscience.org:14001/sparql ?
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/openlink/virtuoso-opensource/issues/131#issuecomment-42976104.

## 

## 

Mark Wilkinson
Madrid, Spain

**Comment by markwilkinson** [comment]
Its the query in the link I posted to the other bug report that is revealing the bug discussed here.  The only difference is that I was requesting JSON output in this case, triggering the bug here.  When requesting NTriples output instead (to get around this bug) I found the other bug that I reported in another thread yesterday.

Apparently, I am cursed to not be able to answer this query ;-)  LOL!

**Comment by haagfn-vis** [comment]
I can reproduce the bug on the aforementioned 3rd party endpoint http://s4.semanticscience.org:14001/sparql with a simpler query such as:

``` sparql
SELECT DISTINCT ?type WHERE 
{ 
  ?s <http://www.openlinksw.com/schemas/virtrdf#qmfName> ?value .  
  FILTER(isLiteral(?value)). 
  BIND(datatype(?value) AS ?type). 
} 
```

This works fine for results formats HTML, Spreadsheet, JavaScript, NTriples, CSV and TSV, but fails with the error message _Virtuoso 22023 Error SR544: Function __xsd_type() can not find XML Schema datatype that matches SQL datatype UNAME (217)_ for results formats XML, JSON and RDF/XML.

The issue can also currently be reproduced the same way on http://dbpedia.org/sparql with the query

``` sparql
SELECT DISTINCT ?type WHERE 
{ 
  ?s <http://dbpedia.org/property/birthdate> ?value .  
  FILTER(isLiteral(?value)). 
  BIND(datatype(?value) AS ?type). 
}
```

where the error occurs for the same formats as listed above, while no error occurs for the others + Turtle.

**Comment by fgiasson** [comment]
This issue is not closed. It has been addressed with the issue #247 which is now fully operational.
